### PR TITLE
fix: periodically refresh market cache

### LINKED ISSUE #7399: DataNode is not in pair with others after restarting from NetworkHistory
# Problem encountered
Background in Slack thread: https://vegaprotocol.slack.com/archives/CEE07A084/p1674237050353779

We restarted `data-node` using NetworkHistory (after `unsafe-reset-all` and dropping the Data-Node database) on Fairground. But vaguer flagged the issues with restarted servers: first `n11`, then `n08`. Our workaround was to restart the nodes from Block 0, and let them replay all blocks.

Issue reported by vaguer:
```
Failed to fetch https://api.n08.testnet.vega.xyz/graphql ([{"message":"no market data for market: aecd0f8a39475cd8432343a083fb6180f0dbcd46c5e584064c5bc1affa868664","path":["marketsConnection","edges",4,"node","data"],"extensions":{"code":13,"type":"Internal"}}])
```

# Observed behaviour
vaguer was reporting node as not in sync with other nodes, after restarting it from NetworkHistory on the Fairground network.

# Expected behaviour
vaguer should give the same score to DataNode started from block zero and DataNode restarted from NetworkHistory.

# Steps to reproduce

## Manual
Open GraphQL Playground, e.g. https://api.n08.testnet.vega.xyz/graphql/ and run the query (from vaguer):
```GraphQL
{
  marketsConnection {
    edges {
      node {
        id
        state
        tradingMode
        accountsConnection {
          edges {
            node {
              type
              balance              
            }
          }
        }
        data {
          auctionStart
          trigger
          markPrice
        }
      }
    }
  }
}
```

Now it gives the result:
```json
{
  "errors": [
    {
      "message": "no market data for market: aecd0f8a39475cd8432343a083fb6180f0dbcd46c5e584064c5bc1affa868664",
      "path": [
        "marketsConnection",
        "edges",
        4,
        "node",
        "data"
      ],
      "extensions": {
        "code": 13,
        "type": "Internal"
      }
    }
  ],
  "data": {
    "marketsConnection": {
      "edges": [
        {
          "node": {
            "id": "8b4aaea9cf7cbbeee90aa6ccea21bd5b9ec15c1872d6b0b3a58e31b741dd948a",
            "state": "STATE_ACTIVE",
            "tradingMode": "TRADING_MODE_CONTINUOUS",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": null,
              "trigger": "AUCTION_TRIGGER_UNSPECIFIED",
              "markPrice": "7367"
            }
          }
        },
        {
          "node": {
            "id": "325dfa07e1be5192376616241d23b4d71740fe712e298130bfd35d27738f1ce4",
            "state": "STATE_ACTIVE",
            "tradingMode": "TRADING_MODE_CONTINUOUS",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "1"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": null,
              "trigger": "AUCTION_TRIGGER_UNSPECIFIED",
              "markPrice": "650900"
            }
          }
        },
        {
          "node": {
            "id": "ad2e531441c2e8a43e85423db399a4acc8f9a8a2376304a4c377d0da8eb31e80",
            "state": "STATE_ACTIVE",
            "tradingMode": "TRADING_MODE_CONTINUOUS",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": null,
              "trigger": "AUCTION_TRIGGER_UNSPECIFIED",
              "markPrice": "2225182214"
            }
          }
        },
        {
          "node": {
            "id": "209e3226375923d461dca53c7cc612f54149586cb06638729c449be20b4aa022",
            "state": "STATE_ACTIVE",
            "tradingMode": "TRADING_MODE_CONTINUOUS",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": null,
              "trigger": "AUCTION_TRIGGER_UNSPECIFIED",
              "markPrice": "164104"
            }
          }
        },
        {
          "node": {
            "id": "aecd0f8a39475cd8432343a083fb6180f0dbcd46c5e584064c5bc1affa868664",
            "state": "STATE_SETTLED",
            "tradingMode": "TRADING_MODE_NO_TRADING",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": null
          }
        },
        {
          "node": {
            "id": "90e71c52b2f40db78efc24abe4217382993868cd24e45b3dd17147be4afaf884",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "c80c1cc89dcc5a302e443b7fcd215cdadbb259f6b180db6253a5572ad6406253",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "acff1b95980c3355e65c7162c35917388702525c0475ba753f03d0efc9c6d6a3",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "06cc8960d987cd4c7ba1364b7d849e5b92fd70267d2d1d91896b418951fe94a4",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "de48ad0adc668a80a074e790d118b6a64fa4909fc60ebbc2c335be3b675fec93",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "41357a60fbb2aa09832742f07012250b3da4d87c38aa0f2e6bbd1391aec10d9f",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        },
        {
          "node": {
            "id": "08500fa5bacb8b381c95524d171c7146b524d1fbd1012e72f063510f41c614a9",
            "state": "STATE_PENDING",
            "tradingMode": "TRADING_MODE_OPENING_AUCTION",
            "accountsConnection": {
              "edges": [
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_FEES_LIQUIDITY",
                    "balance": "0"
                  }
                },
                {
                  "node": {
                    "type": "ACCOUNT_TYPE_INSURANCE",
                    "balance": "0"
                  }
                }
              ]
            },
            "data": {
              "auctionStart": "2023-01-20T13:56:58.102930838Z",
              "trigger": "AUCTION_TRIGGER_OPENING",
              "markPrice": "0"
            }
          }
        }
      ]
    }
  }
}
```

# Evidence

## Additional context
Add any other context about the problem here including; system version numbers, components affected.

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by pscott31**
Closes #7399 
In a slightly hacky way, to be improved later

### CODE REVIEW

File: datanode/service/market.go [hunk 1]
```diff
@@ -46,6 +50,28 @@ func NewMarkets(store MarketStore, log *logging.Logger) *Markets {
...
+			select {
+			case <-ticker.C:
+				{
+					err := m.refreshCache(ctx)
+					if err != nil {
+						m.log.Warn("error refreshing market cache", logging.Error(err))
+					}
+				}
+			}
+		}
```
Comment (jeremyletang) [inline datanode/service/market.go:68]: Could be simplified a little.
```suggestion
		for range time.NewTicker(refreshCacheEvery).C {
			if err := m.refreshCache(ctx); err != nil {
				m.log.Warn("error refreshing market cache", logging.Error(err))
			}
		}
```

**Comment by wwestgarth** [comment]
I think we need the same for the cache in `market_data.go` as well? And also what about `market_depth.go`, that also has an intialise step that I assume also happens pre-network-history loading, is there anything we need to do there?

File: datanode/service/market.go (hunk 1 above)
Comment (pscott31) [reply datanode/service/market.go:68]: Roger wilco

Comment (pscott31) [reply datanode/service/market.go]: (this is done)

**Comment by pscott31** [comment]
> I think we need the same for the cache in `market_data.go` as well? And also what about `market_depth.go`, that also has an `Intialise()` step that I assume also happens pre-network-history loading, is there anything we need to do there?

Well market_data is (at least for active markets) continually updating, so the problem would only affect markets where nothing is going on but you're right. urgh.

I think market depth we're OK because it only makes use of the 'live orders' table which is populated completely from the 'current state' snapshot; e.g. history loading doesn't affect it.

I'll see if I can come up with a better solution whereby we can signal all the stores/services to reinitialise themselves if they want.. but I think if we're want a release in short order this patch is probably OK-enough

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**